src/main/java/org/xbill/DNS/NioTcpClient.java

```diff
@@ -24,16 +25,15 @@
 @Slf4j
 @UtilityClass
 final class NioTcpClient extends Client {
+  private static final AtomicBoolean run = new AtomicBoolean(false);
   private static Queue<ChannelState> registrationQueue;
   private static Map<ChannelKey, ChannelState> channelMap;
-  private static volatile boolean run;
 
   private static void startTcp() throws IOException {
-    if (run) {
+    if (!run.compareAndSet(false, true)) {
```

Comment:
This type of `AtomicBoolean` check is only threadsafe if there is nowhere to set it back to `true`. By setting it back to true in `closeTcp` the atomicity guarantees are removed as someone could call `closeTcp` while this method is still executing (mutating is variables), and that would also allow someone else to get past this point in `startTcp` again.

Reply:
closeTcp isn't actually called anywhere, but yes you're right. The issue that is still left is that another thread would assume the initialization is finished and could access not initialized variables. I'll work on that during the weekend.